### DISCUSSION TITLE
NumberField - too late onChange call

### DISCUSSION THREAD

**Discussion by tomekancu** [adobe/react-spectrum#6061] [Q&A]
Hi, according to the [documentation](https://react-spectrum.adobe.com/react-aria/NumberField.html#controlled) of NumberField the `onChange` handler is only called in the `onBlur` event. 
This leads to unwanted behaviour in the form with appearing fields. 

For [example](https://codesandbox.io/p/sandbox/combobox-blur-problem-forked-4ntv3h) we could have a form that displays an additional field after checking the value of our NumberField. 
In such a view, the user can press the next button without even seeing the additional checkbox. 

This problem wouldn't occur if onChange was called on each key event. 

I thought about a possible solution and I came up with something that doesn't need to change the current cursor position. 

1. For uncontrolled components the solution is quite simple - just call onChange with the current value of numberValue from state - we don't have to worry about the formatting the value because this can  still be done on the blur event.
2. The main problem is for the controlled components, when the value changes we should: 
   - when `value === state.numberValue` - do nothing, not even formatting to keep the current cursor position, the formatting can be done on blur 
   - when `value !== state.numberValue` - set the numberValue and move the cursor to the end of the text(the controlled value is the most important) 

To make sure this works, numberValue should have a value that would be emitted with onBlur for the  current inputValue state.  

What do you think about this solution?

**Comment by snowystinger** [comment]
Thanks for your thoughts. This is definitely a tricky situation.

I think this runs into an issue where the value from onChange may not match the value the user thinks they entered. For instance, if the value entered is not parseable to a number, then some nullish value would be in the onChange and it'd only be revealed onBlur when the formatter fails or parses to something else.

Instead, I'd consider always showing your fields so that users don't accidentally skip them. Users can also be prepared for all the fields they may encounter.

Finally, I think hiding and showing form fields can be problematic. There wouldn't be an announcement for a screen reader necessarily for something like this. You could validate the age on submit and create an error on that field. I would also recommend a description on the numberfield which tells users the acceptable range. You could also make use of the min/max value.

**Reply by tomekancu** [reply]
> For instance, if the value entered is not parseable to a number, then some nullish value would be in the onChange and it'd only be revealed onBlur when the formatter fails or parses to something else.

I'm not sure this is the case. I think it is rather unlikely. 
NumberField only allows to type only digits, thousands separators, and decimal separators.
* Additional thousands separators or white spaces are already skipped
* You cannot write more than one decimal separator
To return some a nulish value in onChange, you need to clear the whole field.  

Even then, to avoid clearing the controlled field, we can skip such onChange. 
This is how the dateField's onChange work right now. The onChange is called as soon as the valid date is entered. 

> Instead, I'd consider always showing your fields so that users don't accidentally skip them. Users can also be prepared for all the fields they may encounter.

Here is the problem. We don't want to show fields that are not needed at the moment to not scare the user - according the UX team. 
Given [example](https://codesandbox.io/p/sandbox/combobox-blur-problem-forked-4ntv3h) is very simple, only one field appears, but in some scenario it could be the whole page. 

> There wouldn't be an announcement for a screen reader necessarily for something like this.

I'm not sure if it's needed when the field would appear during every key event. Pressing tab just comes to the next showed field. 

> You could validate the age on submit and create an error on that field. I would also recommend a description on the numberfield which tells users the acceptable range. You could also make use of the min/max value.

This is not possible, the real case is much more complicated.

**Reply by snowystinger** [reply]
> I'm not sure this is the case. I think it is rather unlikely.
NumberField only allows to type only digits, thousands separators, and decimal separators.
>
> Additional thousands separators or white spaces are already skipped
You cannot write more than one decimal separator
To return some a nulish value in onChange, you need to clear the whole field.

You can type other characters, such as currency symbols, negative and positive symbols, parenthesis for accounting negative numbers, numbers in other numbering systems such as Arabic, group and decimal separators which may differ across locales, and a whole host of others.

Right now we will prevent you from typing something which doesn't parse, however, that does not mean the thing we are parsing to matches the value the user expects. We'd love to do real time formatting, but we aren't there yet. See https://github.com/adobe/react-spectrum/issues/1879

For instance, '1.250' could be one point two five to an American, but if it's a German locale website, it could be one thousand two hundred fifty. This is why settings like min/max/significant digits/fraction digits are all important.

Dates are actually easier, the format is known ahead of time, as well as the numeral system.

@majornista for any a11y advice

**Reply by tomekancu** [reply]
@snowystinger I was able to prepare a solution based on my idea. Here is the [codesandbox](https://codesandbox.io/p/sandbox/numberfield-onchange-controlled-form-forked-53sp99). 
I think it works quite well. Even for less standard locales it works as well as the original. I haven't noticed the problems you mention. 

At the moment it is quite hacky to get access to the internal state of NumberField, but I think it could be implemented with success directly in `useNumberFieldState` and `useNumberField`.

Might be worth trying. The current behaviour of onChange may be problematic.

**Reply by snowystinger** [reply]
Thanks so much for the sandbox; I'm impressed; it appears to be working. I'm bringing it up with the team for more opinions.

**Reply by snowystinger** [reply]
I was playing around with this some more today, and I noticed an issue. [sandbox](https://codesandbox.io/p/sandbox/numberfield-with-custom-onchange-forked-ly2krw?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cltt3zi7m00063b6i6t757tg3%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cltt3zi7l00023b6ifx3qguh6%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cltt3zi7l00033b6ittzj9xox%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cltt3zi7l00053b6i9r4vkb2k%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cltt3zi7l00023b6ifx3qguh6%2522%253A%257B%2522id%2522%253A%2522cltt3zi7l00023b6ifx3qguh6%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522cltt3zi7l00053b6i9r4vkb2k%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cltt3zi7l00043b6ib87x6z6w%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cltt3zi7l00053b6i9r4vkb2k%2522%252C%2522activeTabId%2522%253A%2522cltt3zi7l00043b6ib87x6z6w%2522%257D%252C%2522cltt3zi7l00033b6ittzj9xox%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cltt3zi7l00033b6ittzj9xox%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

Type 200 in the controlled input, then type 100, the field does not format again.

In addition, I've remembered one of the flows which initially caused us to not implement this. Using the same sandbox, refreshed, type '35', then move the cursor in between the 3 and 5, type '0'. Cursor position is lost.

**Reply by tomekancu** [reply]
Yes, it has some drawbacks. In your example it could be solved by using `maxValue` than a custom `setV` function. 
My solution has an assumption, that the controlled value should not be modified while the user is typing. Even in the native NumberField changing the controlled value while the user is typing would do that. 
Such modification should be done in `onBlur`.

**Reply by snowystinger** [reply]
Yes, my example was contrived, but that was for simplicity. You could be avoiding prime numbers or something else which can't be handled via min/max. While I would say handle that in validation, the temptation is there. It's easy to reach for.

I would say that we are in agreement that the controlled value should not be modified while the user is typing. We are also currently enforcing this.

If we open it up, this is an easy rule to break, leading to the bad behaviors I've pointed out. I don't think people would do it on purpose; it'd just be hard to do correctly.

**Reply by tomekancu** [reply]
I agree that this change would lead to hidden bugs. On the other hand, this behaviour is also present in React's native input. 
Adding such complicated logic in onChange for `ev.target.value` would  lead to the same behaviour. 
This suggests that this is a misuse of `onChange`.

I can imagine that changing the behaviour of `onChange` now would be a breaking change for someone with the validation you describe.
Perhaps the best approach would be to add a second event handler `onValueInput`(fired on every change) to allow implementations from my examples. 
This will force developers to think twice and make the change intentionally. 

I believe it is worth trying.
As seen in my example, it is not that unusual for some fields to be dependent from the others. 
The problem may also occur with the disable/readonly attribute - it is not just a problem with hidden fields.

**Reply by snowystinger** [reply]
> this behaviour is also present in React's native input

I don't think I follow. I wouldn't try to align React's goals with our goals though. They aim to allow developers to build components. We aim to help developers build accessible applications. As long as a user can meet the controlled/uncontrolled paradigm requirements, I don't see any misuse of `onChange`. There are many components which have these two props (or more pairs) and have no native input backing them. That this one happens to use an input is an implementation detail.

The `onValueInput` has some merit, and I was actually just discussing it here: https://github.com/adobe/react-spectrum/issues/6156#issuecomment-2041646952
I still have worries there about misuse, though restricting it to the event only and not the value is an interesting idea. However, I'm still worried people would try to use it to do their own parsing and feed it back into the `value` prop or rely on the partial parsed value which may be incorrect. See link for why I think that is a concern.

Another option available to you is to set `autoFocus` or manage the focus yourself to direct the user around these fields which can appear and disappear.

**Reply by tomekancu** [reply]
> this behaviour is also present in React's native input

What I meant was that using `onChange` in a simple `<input />` could cause the same problems with cursor, as you did with your setV function.

> As long as a user can meet the controlled/uncontrolled paradigm requirements, I don't see any misuse of onChange. 

Keeping example with just an `<input />`.  Let's assume that someone want's to trim spaces from ends of the text(like you did with keeping `value <= 50`).
```jsx
function Component() {
  const [v, setV] = React.setState("");
  const handleChange = ev => setV(ev.target.value.trim());

  return <input value={ v } onChange={ handleChange } />
}
```
 Such an implementation would prevent the user from using spaces in the middle of the text while typing. 
That's what I meant by misusing of `onChange` and React's native behaviour of onChange(you should not change value during onChange)

**Reply by snowystinger** [reply]
> That's what I meant by misusing of onChange and React's native behaviour of onChange (you should not change value during onChange)

Right, so we're saying the same thing again.

I see a couple issues with using:

`onInputValue: (val: number) => void`
This can result in a val of NaN depending on partial parsing, or it might just be another value completely.

`onInputValue: (val: string) => void`
The user would need to parse this, which would be error prone due to the locales and numbering systems we support.

In addition, both could be misused for a controlled state and opens us to bugs if there are two sources of truth to control the component.

I did have another thought, though. If the Tab key is pressed, we could possibly commit during that, much like we do for the 'Return' key, and get ahead of the blur. Do you think that would work for your use case? I haven't thought through the implications of this fully yet.

**Reply by tomekancu** [reply]
> This can result in a val of NaN depending on partial parsing, or it might just be another value completely.

I cannot reproduce the case. In all my tests, the number was parsed correctly - even in different numeric systems. 
I was only allowed to use allowed characters(thousand separator, decimal separator, number), additional thousand separators or decimal separators were skipped during parsing. 
The only situation where I got NaN, was an empty string.

>  If the Tab key is pressed, we could possibly commit during

This will help a little bit, but only for users who only use the keyboard to move through the form.

**Reply by snowystinger** [reply]
The failed parsing usually comes in with partial currency or units, you can check out our tests here https://github.com/adobe/react-spectrum/blob/b2d25ef23b827ec2427bf47b343e6dbd66326ed3/packages/%40internationalized/number/test/NumberParser.test.js#L105

Admittedly, it is hard to type out the currency/units right now, but that's something we'd like to address. https://github.com/adobe/react-spectrum/issues/2806

**Reply by tomekancu** [reply]
If I understand correctly:
When you fix the issue #2806, users will be able to delete part of the currency sign producing NaNs, but they won't be able to write the currency by themselves.
Users still won't be able to write other characters by themselves - only delete them.

Without the fix it is hard to say how my solution would work. Maybe skipping this extra onChange for NaNs would be enough.
